app-libs/stf/src/getter.rs

```diff
@@ -0,0 +1,198 @@
+	limitations under the License.
+
+*/
+
+use crate::{AccountId, KeyPair, Signature};
+use codec::{Decode, Encode};
+use itp_stf_interface::ExecuteGetter;
+use log::*;
+use sp_runtime::traits::Verify;
+use std::prelude::v1::*;
+
+#[cfg(feature = "evm")]
+use sp_core::{H160, H256};
+
+#[cfg(feature = "sgx")]
+pub use sgx::*;
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[allow(non_camel_case_types)]
+pub enum Getter {
+	public(PublicGetter),
+	trusted(TrustedGetterSigned),
+}
+
+impl From<PublicGetter> for Getter {
+	fn from(item: PublicGetter) -> Self {
+		Getter::public(item)
+	}
+}
+
+impl From<TrustedGetterSigned> for Getter {
+	fn from(item: TrustedGetterSigned) -> Self {
+		Getter::trusted(item)
+	}
+}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[allow(non_camel_case_types)]
+pub enum PublicGetter {
+	some_value,
+}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[allow(non_camel_case_types)]
+pub enum TrustedGetter {
+	free_balance(AccountId),
+	reserved_balance(AccountId),
+	nonce(AccountId),
+	#[cfg(feature = "evm")]
+	evm_nonce(AccountId),
+	#[cfg(feature = "evm")]
+	evm_account_codes(AccountId, H160),
+	#[cfg(feature = "evm")]
+	evm_account_storages(AccountId, H160, H256),
+}
+
+impl TrustedGetter {
+	pub fn sender_account(&self) -> &AccountId {
+		match self {
+			TrustedGetter::free_balance(sender_account) => sender_account,
+			TrustedGetter::reserved_balance(sender_account) => sender_account,
+			TrustedGetter::nonce(sender_account) => sender_account,
+			#[cfg(feature = "evm")]
+			TrustedGetter::evm_nonce(sender_account) => sender_account,
+			#[cfg(feature = "evm")]
+			TrustedGetter::evm_account_codes(sender_account, _) => sender_account,
+			#[cfg(feature = "evm")]
+			TrustedGetter::evm_account_storages(sender_account, ..) => sender_account,
+		}
+	}
+
+	pub fn sign(&self, pair: &KeyPair) -> TrustedGetterSigned {
+		let signature = pair.sign(self.encode().as_slice());
+		TrustedGetterSigned { getter: self.clone(), signature }
+	}
+}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct TrustedGetterSigned {
+	pub getter: TrustedGetter,
+	pub signature: Signature,
+}
+
+impl TrustedGetterSigned {
+	pub fn new(getter: TrustedGetter, signature: Signature) -> Self {
+		TrustedGetterSigned { getter, signature }
+	}
+
+	pub fn verify_signature(&self) -> bool {
+		self.signature
+			.verify(self.getter.encode().as_slice(), self.getter.sender_account())
+	}
+}
+
+#[cfg(feature = "sgx")]
+pub mod sgx {
+	use super::*;
+	use ita_sgx_runtime::System;
+	use itp_utils::stringify::account_id_to_string;
+
+	#[cfg(feature = "evm")]
+	use ita_sgx_runtime::{AddressMapping, HashedAddressMapping};
+
+	#[cfg(feature = "evm")]
+	use crate::evm_helpers::{get_evm_account, get_evm_account_codes, get_evm_account_storages};
+
+	impl ExecuteGetter for Getter {
+		fn execute(self) -> Option<Vec<u8>> {
+			match self {
+				Getter::trusted(g) => match &g.getter {
```

Comment:
I totally agree, but I don't want to introduce too many changes in one PR. And restructuring / removing the whole GetterExecutor should be a separate step, I believe.

But I agree about making the Runtime generic. I'll try that.

Reply:
It's that simple, because we have hardcoded Getter and Call values. So even though the Runtime would be generic implemented like this, the Runtime parameters are not. 

So we would have to make the Getter and Call parameter Runtime dependent as well. I'm not sure how much that would make sense, because the enum (e.g. `TrustedGetter::free_balance` does not make sense without a balance pallet anyway) itself needs to be adapted when the Runtime is changed.

For now, I was able to make the Stf itself generic over the runtime, because the Getters and Calls are abstracted over the ExecuteCall and ExecuteGetter Trait. If the Runtime is changed, one can simply inject another Getter / Call Type.

Reply:
Yes, we would not have completely generified the whole thing then. However, now we are forced to use the `sgx_runtime`, which does not compile to `std` if I am not mistaken. If we made it generic over the runtime, we could use any runtime in the tests.

> For now, I was able to make the Stf itself generic over the runtime, because the Getters and Calls are abstracted over the ExecuteCall and ExecuteGetter Trait. If the Runtime is changed, one can simply inject another Getter / Call Type.

This already seems like a very valuable addition 👍.

I am satisfied for now with the current approach for now! Thanks for tackling this.

Reply:
sgx-runtime is actually available in std. 🥳

Reply:
Proposing a follow up issue with the following content: 

> Abstract Getter and Calls over the Runtime. This would need some refactoring of the currently hardcoded enums. This could be an extenable Getter struct / enum. E.g. A Balance-Pallet-Getter implentation which is added to the general Getter implementation iff needed for the current Runtime.

Reply:
I am ok with that. However, I am not yet convinced that we need the Getter/Calls struct, I think `ExecuteGetter/-Calls` traits could be implemented on the STF directly, which would then be abstracted over the runtime.

Reply:
https://github.com/integritee-network/worker/issues/1037